README.md
Atomic Lacuna
=============

 > The Lacuna Expanse Web Client inside an atom-shell. It's cleaner, sweeter, brighter
 > and includes a scripting console.

In terms of documentation, the isn't much, *yet*. But hey,
if you want the short and sweet of it, see the ROADMAP. Or for more detailed (read: technical)
documentation see the GOALS. Finally if you're a developer and want to help, CHANGES
and HACKING are the places to be.

If you have any questions or suggestions, please yell at me via the Github Issue tracker.
I'd love to hear from you. :blush:

Why? :octocat:
====

I've been wanting to try and add cool new stuff to the Lacuna Web Client for
quite some time now, however, my plans have been foiled by the inability to load
more JavaScript into the page (Underscore or jQuery to name a few). This was
because the `index.html` couldn't be changed and I didn't want to be inserting
`<script>` elements onto the page at runtime ('cause that's ugly, man).

I came across [Atom](atom.io) some time ago, however I never really looked into it.
Recently I tried to build it on my system and I discovered that it ran on a 'shell'
that could be re-used in other projects.

To make a short story shorter, the `atom-shell` allows a web page (such as the
web client) to run in a desktop window and have access to the file system via
NodeJs. This allowed me to streamline the loading of all the JavaScript and assets.
Several users who have tested Atomic Lacuna have said that it feels faster than
the in-browser client. Here's a sneak peak of what it looks like. Tadaaa! :dizzy:

![Atomic Lacuna in action!](https://raw.githubusercontent.com/Vasari/atomic-lacuna/master/public/Screenshot.png)

Badges! :beginner:
=======

| Name             | Status    |
| :--------------: | :-------: |
| Code Climate     | [![Code Climate](https://codeclimate.com/github/Vasari/atomic-lacuna.png)](https://codeclimate.com/github/Vasari/atomic-lacuna) |
| Travis CI        | [![Build Status](https://travis-ci.org/Vasari/atomic-lacuna.svg?branch=master)](https://travis-ci.org/Vasari/atomic-lacuna)     |
| Ohloh            | [![Ohloh](https://www.ohloh.net/images/logo/ohloh_hub.png)](https://www.ohloh.net/p/atomic-lacuna)
| Dependencies     | [![Dependency Status](https://david-dm.org/Vasari/atomic-lacuna.svg)](https://david-dm.org/Vasari/atomic-lacuna) |
| Dev Dependencies | [![Dev Dependency Status](https://david-dm.org/Vasari/atomic-lacuna/dev-status.svg)](https://david-dm.org/Vasari/atomic-lacuna#info=devDependencies) |

Made with :heart: by Nathan McCallum under the MIT license! :panda_face:


CHANGES.md
Changes
=======

 > History of the changes made to the code since version `0.0.1`.

###0.3.0 (Unreleased)
- Massive overhaul of the documentation.

###0.2.0 (2014-07-26)
- Cleaned all of the code out using a script.
- Fixed some errors introduced in the cleaning.
- Implemented a build for the web browser.
- Various improvements to the internals of the code.

###0.1.1 (2014-07-13)
- Include binaries for Windows (32 bit) and Linux (64 and 32 bit).

###0.1.0 (2014-07-13)
- Started keeping a change log.
- Sorted out the inclusion of `node_modules` in linux builds.
- The build script now zips each package.
- Build script now deletes old stuff before running.
- Ignore useless development dependencies when building.

###0.0.1
- Initial version! :D


build.rb
# encoding: utf-8


####
#
# The build script for all this crazy stuff.
#
####

require 'fileutils'
require 'json'

# These are non-core dependencies. If you're getting an error on this line, you'll
# need to get and run bundler in this directory. (Google it. :P)
require 'archive/zip'
require 'handlebars'

# Auto-Flush STDOUT
STDOUT.sync = true


# Get our working area sorted out.
build_dir = File.join(Dir.getwd, 'build')
FileUtils::mkdir_p build_dir


PACKAGES = %w(web linux-32 linux-64 win32-32)

# Clean out the old build files 'n' things.
puts 'Cleaning out files from previous run...'
(PACKAGES + ['app']).each do |package|
    bin = File.join(build_dir, package)
    zip = bin + '.zip'

    FileUtils.remove_entry_secure bin if Dir.exists? bin
    FileUtils.remove_entry_secure zip if File.exists? zip
end

print "\n\n"



# Run the Gulp task which downloads the atom-shell for different platforms.
unless system 'gulp download-shell'
    puts "ERR: Failed to download atom-shell"
    exit
else
    print "\n\n"
end


# Do this first, because it's needed for every build type.
puts "Running Browserify on the JavaScript code."
puts "Concatenating all the CSS files."
successful = system 'gulp code-build'
print "\n\n" # some space

if successful

    puts "Pulling all the code together."
    FileUtils.mkdir_p File.join(build_dir, 'app')
    FileUtils.cp('public/dist/browser-build.js', 'build/app/browser-build.js')
    FileUtils.cp('public/dist/templates.js', 'build/app/templates.js')
    FileUtils.cp('public/dist/styles.css', 'build/app/styles.css')
    FileUtils.cp('public/index-template.html', 'build/app/index-template.html')
    FileUtils.cp('public/browser.html', 'build/app/browser.html')
    FileUtils.cp('src/main.js', 'build/app/main.js')

    # Clean up the package.json file so that the 'main' attribute points correctly to
    # the app's launch file. (Which has obviously changed...)
    json = JSON.parse(File.read('package.json'))
    json['dependencies'] = nil
    json['devDependencies'] = nil
    json['main'] = 'main.js'

    # Reject all that are nil so that we clear the keys out as well.
    json.reject! { |k, v| v.nil? }

    # Finally write the json back to where we want it.
    File.write('build/app/package.json', JSON.pretty_generate(json, {indent: ' ' * 4}))
else
    puts 'ERR: code-build failed. Die die die!'
    exit
end


# Finally, let's get started.
print "\n\n" # some space
PACKAGES.each do |build|
    bin_path = File.join(build_dir, build)

    puts "Constructing the #{build} package."


    if build == 'web'
        # In this case, bin_path is the location we're copying all this stuff to.
        src = File.join(build_dir, 'app')
        dest = bin_path
        FileUtils.mkdir_p dest

        # Compile the HTML template for the index page.
        handlebars = Handlebars::Context.new
        template = handlebars.compile(File.read(File.join(src, 'index-template.html')))
        html = template.call(atomShell: false, development: false)
        File.write(File.join(dest, 'index.html'), html)

        # Copy all assets needed for the browser.
        FileUtils.cp(File.join(src, 'styles.css'), dest)
        FileUtils.cp(File.join(src, 'browser-build.js'), dest)
        FileUtils.cp(File.join(src, 'templates.js'), dest)
    else

        unless Dir.exists? bin_path
            puts "ERR: please download the #{build} binary."
            next
        end

        src = File.join(build_dir, 'app')
        dest = File.join(bin_path, 'resources', 'app')
        FileUtils.mkdir_p dest # make sure we have somewhere to go..

        # Move the templates and JavaScript into place.
        FileUtils.cp_r('app/js/.', File.join(dest, 'js'))
        FileUtils.cp_r('app/templates/.', File.join(dest, 'templates'))
        FileUtils.cp_r('app/lib/.', File.join(dest, 'lib'))

        # Woo! resources.json!
        FileUtils.mkdir_p File.join(dest, 'data')
        FileUtils.cp_r('app/data/resources.json', File.join(dest, 'data', 'resources.json'))

        # Need to copy node_modules because this code isn't run through Browserify.
        # So, grab a list of dependencies (not devDependencies) and copy them across.
        # Ignoring devDependencies is important because it reduces the size
        # down quite a bit. (Browserify is 25 mb!)
        JSON.parse(File.read('package.json'))['dependencies'].each do |name, ver|
            from = File.join('node_modules', name) + '/.'
            to = File.join(bin_path, 'node_modules', name)

            # Make sure we have somewhere to go.
            FileUtils.mkdir_p to
            FileUtils.cp_r(from, to)
        end

        FileUtils.cp_r([
            File.join(src, 'package.json'),
            File.join(src, 'styles.css'),
            File.join(src, 'index-template.html'),
            File.join(src, 'main.js'),
        ], dest)
    end

    # Now we need to zip this thing!!
    Archive::Zip.archive(File.join(build_dir, build + '.zip'), bin_path)
end

print "\n\n"
puts "All finished! Ready to publish on Github! :D"
print "\n"
